src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,175 @@
+                            peReference,
+                            handleAccessibleItem,
+                            cancellationToken);
+                    }
+                }
+            }
+
+            return;
+
+            static void AddItems(CompletionItem item, CompletionContext completionContext, ImmutableHashSet<string> namespacesInScope)
+            {
+                var containingNamespace = TypeImportCompletionItem.GetContainingNamespace(item);
+                if (!namespacesInScope.Contains(containingNamespace))
+                {
+                    completionContext.AddItem(item);
+                }
+            }
+        }
+
+        private ImmutableHashSet<string> GetNamespacesInScope(Document document, SyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            // This hashset will be used to match namespace names, so it must have the same case-sensitivity as the source language. 
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var builder = ImmutableHashSet.CreateBuilder(syntaxFacts.StringComparer);
+
+            var semanticModel = syntaxContext.SemanticModel;
+            GetImportedNamespaces(syntaxContext.LeftToken.Parent, semanticModel, builder, cancellationToken);
+
+            // Get containing namespaces.
+            var namespaceSymbol = semanticModel.GetEnclosingNamespace(syntaxContext.Position, cancellationToken);
+            while (namespaceSymbol != null)
+            {
+                builder.Add(namespaceSymbol.ToDisplayString(SymbolDisplayFormats.NameFormat));
+                namespaceSymbol = namespaceSymbol.ContainingNamespace;
+            }
+
+            return builder.ToImmutable();
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = default, CancellationToken cancellationToken = default)
+        {
+            var newDocument = await ComputeNewDocumentAsync(document, item, cancellationToken).ConfigureAwait(false);
+            var newText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var changes = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(newText, changes.ToImmutableArray());
+
+            return CompletionChange.Create(change);
+        }
+
+        private async Task<Document> ComputeNewDocumentAsync(Document document, CompletionItem completionItem, CancellationToken cancellationToken)
+        {
+            var containingNamespace = TypeImportCompletionItem.GetContainingNamespace(completionItem);
+            Debug.Assert(containingNamespace != null);
+
+            // Complete type name.
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textWithTypeName = text.Replace(completionItem.Span, completionItem.DisplayText);
```

Comment:
note: you may need some tests where the completion item is hte name of some keyword.  more likely to happen in VB than c#.    there may be a bug here (though it also might not be important enough to fix).